pyqtorch/core/utils.py

```diff
@@ -86,23 +85,20 @@ def _apply_gate(
     - state (torch.Tensor): the quantum state after application of the gate.
     Same shape as `ìnput_state`
     """
-    mat = mat.reshape([2] * len(qubits) * 2)
+    mat = mat.view([2] * len(qubits) * 2)
     mat_dims = list(range(len(qubits), 2 * len(qubits)))
-    state_dims = [N_qubits - i - 1 for i in list(qubits)]
-    axes = (mat_dims, state_dims)
-
+    qubits = list(qubits)
+    axes = (mat_dims, qubits)
     state = torch.tensordot(mat, state, dims=axes)
     inv_perm = torch.argsort(
-        torch.tensor(
-            state_dims + [j for j in range(N_qubits + 1) if j not in state_dims], dtype=torch.int
-        )
+        torch.tensor(qubits + [j for j in range(N_qubits + 1) if j not in qubits], dtype=torch.int)
     )
     state = torch.permute(state, tuple(inv_perm))
```

Comment:
What is this permutation doing?

Reply:
when you do `tensordot` on a subset of dims on two tensors, the contracted dimensions are added as the first dimensions of the tensor, the permute reverts this back to the original order